### PR TITLE
Improve "Open in New Window" feature

### PR DESCRIPTION
**Pull request by codebrainz**
Make an attempt to use the same executable file for the new window as the current instance of Geany. Also flush configuration to disk and provide path to it for the new instance using the `-c` option.

Not mentioned in the commit message, I also changed the error cases to use `ui_set_statusbar()` instead of `g_printerr()` to give a little better feedback to the user if it fails.

<del>The extra steps to find the correct Geany executable should fall back to the existing behaviour if they fail. <del>There's some [additional platform-specific ways to find the executable path](http://stackoverflow.com/a/1024937) if someone feels like implementing them.</del>(done)</del>(people didn't like I guess).

<del>The Win32 code path is not tested but "should work" (famous last words) barring any typos.</del>(uses same code as other platforms now).

<del>The OSX bundle case is not handled yet, see the `TODO:` comments left in the source.</del>

<del>I updated my branch. As before it still tries the platform-specific ways to get the executable path first, and then falls back to other ways if they fail.</del>

<del>Currently supports:</del>
- <del>Linux</del>
- <del>Windows</del>
- <del>OSX with or without bundle (untested)</del>
- <del>FreeBSD, NetBSD, DragonFly BSD (untested)</del>
- <del>Every other platform Geany supports</del>
  - <del>If Geany is installed, it should be correct executable</del>
  - <del>If not, it will be correct executable if same `geany` is in the `PATH` (as before)</del>

An additional change added, as with PR #446 it tries the actual installation path if the platform-specific way failed, before finally falling back to the existing `g_find_program_in_path()`. Unlike PR #446, it tries to use the same configuration directory by passing `-c` argument.

Unlike PR #606, it doesn't try to use the program name from `argv[0]` because it's not really a great candidate as it might not even contain a path, let alone the path to the actual executable, and also I didn't really understand the logic in that PR with respect to trying to extract a meaningful path from it. This could be added as a final fallback before trying `g_find_program_in_path()` if people really want it.

<del>It's kind of ugly using platform-specific code like this, but this really is a platform-specific thing trying to be determined and all of that code is nicely isolated inside a single trivial function (instead of spread around more as with PR #606). If people really don't like the platform-specific code, we could remove the roughly 60-70 lines of commented `#ifdef` stuff in `utils_find_self_executable()` and it would be pretty much like an improved version of PR #446.</del>(done)

I should also note that it's still using `utils_spawn_async()` instead of the new spawn functions, just because it's kind of unrelated to the PR. If anyone wants to use other functions, we could add that as a separate commit.

### CODE REVIEW

**Comment by codebrainz** [comment]
@techee for the OSX bundle, I expect if `_NSGetExecutablePath()` was used it would give the path to the executable inside the bundle. If that's the case, could we remove the special handling for `is_osx_bundle()` and use the same code path for both? I tried to leave the current code for OSX bundle alone since I wasn't exactly sure how to pass arguments to the bundle through the `open` utility, if it's even possible.

**Comment by codebrainz** [comment]
Fixed the Win32 build, let's see if it makes Travis happy :)

Normally I would `#define WIN32_LEAN_AND_MEAN` before including `windows.h` to avoid including a bunch of unused stuff, but after @zhekov's long explanations, I'm too afraid, so meh :)

**Comment by zhekov** [comment]
@codebrainz 

> Normally I would |#define WIN32_LEAN_AND_MEAN| before including 
> |windows.h| to avoid including a bunch of unused stuff, but after 
> @zhekov <https://github.com/zhekov>'s long explanations, I'm too afraid, 
> so meh :)

Oh... sorry about that. :) If only GetModuleFileName() is what you need, that's part of the most basic WINAPI, so you can define any LEAN you want. Not that it'll have any noticeable effect...

However:

"// Try reading link '/proc/self/exe' which works on <b>some</b> *nix platforms"

A quick search shows [1] that'll probably only work on Linux (specifically, not on OSX). Various *nixes have different names, it's the usual mess.

Personally I would have done the whole thing somewhat different:

1. If argv[0] is absolute, use that.
2. If argv[0] is relative, try to combine it's basename with bindir (extend utils_resource_dir() with bindir first).
3. If both fail, try spawn_async() with argv[0]'s basename (if any) or with "geany". It will always search the path, and has less arguments.

It's not like Geany will run from any directory - it *requires* geany.glade, and that's provided by utils_get_locale_from_utf8(utils_resource_dir(RESOURCE_DIR_DATA)). Adding RESOURCE_DIR_BIN shoudn't be a problem.

The above steps are for *nix/Win~1. I have no idea about OSX, but utils_resource_dir() does indeed support is_osx_bundle()... Trying to find the executable name in an OS-specific manner seems like a lot of trouble.

[1] http://stackoverflow.com/questions/1023306/finding-current-executables-path-without-proc-self-exe

**Comment by codebrainz** [comment]
On 15-08-04 10:41 AM, zhekov wrote:
> @codebrainz
>
>> Normally I would |#define WIN32_LEAN_AND_MEAN| before including
>> |windows.h| to avoid including a bunch of unused stuff, but after
>> @zhekov <https://github.com/zhekov>'s long explanations, I'm too afraid,
>> so meh :)
>
> Oh... sorry about that. :) If only GetModuleFileName() is what you need, that's part of the most basic WINAPI, so you can define any LEAN you want. Not that it'll have any noticeable effect...
>

Yeah, I figured it's no big deal either way.

> However:
>
> "// Try reading link '/proc/self/exe' which works on <b>some</b> *nix platforms"
>
> A quick search shows [1] that'll probably only work on Linux (specifically, not on OSX). Various *nixes have different names, it's the usual mess.
>

Note that I linked that same SO question from the PR description :) 
Basically I provided implementations for Linux and Windows (and left 
notes about the OSX implementation), and mentioned that if anyone else 
wants to provide implementations for other platforms, that link has more 
info, patches welcome.

The goal for me was to open the exact same executable that the current 
instance is on Linux (what I use), and leave the old behaviour for other 
platforms. On Linux /proc/self/exe is exactly the best (and only?) way 
to get the path to the current executable.

> Personally I would have done the whole thing somewhat different:
>
> 1. If argv[0] is absolute, use that.

Argv is kind of a bummer since it can be changed in a number of ways, 
and in this specific case, we actually hard-code it to a non-absolute 
path, so it would never do what I wanted.

> 2. If argv[0] is relative, try to combine it's basename with bindir (extend utils_resource_dir() with bindir first).

Wouldn't work for the OSX bundle case since we hard-code argv[0] to 
"open" (a utility like start.exe on Windows or xdg-open on Linux). It 
will never be in the bundle (if OSX bundle even has a bindir).

> 3. If both fail, try spawn_async() with argv[0]'s basename (if any) or with "geany". It will always search the path, and has less arguments.
>

The old code already searched the path, that's what I was trying to 
improve, and leaving it as the fallback behaviour for other platforms.

> It's not like Geany will run from any directory - it *requires* geany.glade, and that's provided by utils_get_locale_from_utf8(utils_resource_dir(RESOURCE_DIR_DATA)). Adding RESOURCE_DIR_BIN shoudn't be a problem.
>
> The above steps are for *nix/Win~1. I have no idea about OSX, but utils_resource_dir() does indeed support is_osx_bundle()... Trying to find the executable name in an OS-specific manner seems like a lot of trouble.
>

It was actually pretty easy, just a handful of lines for each platform 
anyone cares enough about to support, and leaving the old path-searching 
behaviour to fall through to for other platforms.

**Comment by zhekov** [comment]
>> 1. If argv[0] is absolute, use that.

> Argv is kind of a bummer since it can be changed in a number of ways, and in this specific case, we actually hard-code it to a non-absolute path, so it would never do what I wanted.

I mean the original argv[0] as an argument of Geany's main(); we pass argv to main_lib() directly, and don't "hardcode" argv[0] AFAIK. Same for points 2 and 3.

And if one execs geany, but passes an argv[0] of "foo" as exec()'s argv[0], so be it - we should try to respect that.

It may not work for OSX bundle, but this PR's OSX is not finished either, and the current *nix support is actually Linux-only.

**Comment by codebrainz** [comment]
On 15-08-04 01:08 PM, zhekov wrote:
>>> 1. If argv[0] is absolute, use that.
>
>> Argv is kind of a bummer since it can be changed in a number of ways, and in this specific case, we actually hard-code it to a non-absolute path, so it would never do what I wanted.
>
> I mean the original argv[0] as an argument of Geany's main(); we pass argv to main_lib() directly, and don't "hardcode" argv[0] AFAIK. Same for points 2 and 3.
>

Yeah, that would be somewhat of an improvement if it was an absolute 
path, if not, it would probably be kind of brittle and messy code, 
compared to the old (and "improved") way.

> And if one execs geany, but passes an argv[0] of "foo" as exec()'s argv[0], so be it - we should try to respect that.
>

Which "foo" though? The point of this PR is to try and help it use the 
exact same binary code as the current instance.

In a perfect world, the "Open in New Window" feature would actually 
just open another GtkWindow in the same process, but due to the design 
of Geany, this would most likely be insanely difficult to implement[0].

> It may not work for OSX bundle, but this PR's OSX is not finished either, and the current *nix support is actually Linux-only.
>

Linux is a *nix, and at least a handful of others support the same 
concept just using a different filename, which is why I offered anyone 
using such platforms to feel free to add.

The current PR improves the situation on Linux and Windows, and leaves 
it simple for people using other platforms, who can actually test code 
on them, to add support if they didn't like the old way. Otherwise it 
does exactly what it did before (search PATH).

I even left the exact function name for the OSX implementation in a 
comment, I just didn't want to blindly implement it because I won't be 
able to test it, I'm not familiar with OSX C-runtime, and we don't have 
continuous integration support for OSX, so I won't even know if it 
builds like I do for Windows. Given that this feature previously didn't 
work on OSX[1] and is now a step closer, I'd say that qualifies as an 
improvement :)

----

[0]: Though it would cleanup a lot of code if we did it using the 
provided GTK+ mechanisms for this like GtkApplication and friends. We 
could drop all the socket/instance related code at a minimum, and 
supporting two windows would force a lot of the code to be re-written in 
a more extensible/modular way.

[1]: It fails silently leaving a debug message "Unable to open new 
window: Failed to execute child process "UTF-8" (No such file or 
directory)".

**Comment by codebrainz** [comment]
I'm a little unsure also whether `argv[0]` would even be reliable in all cases. According to the C standard it's up to the implementation and should "... _represent_ the program name". Even according to POSIX, which uses loosey-goosey wording like "... _should_ point to a filename string that is _associated_ with the process being started". Neither seems to require that it holds the name of a/the binary (though it most likely does outside of `exec*()` functions), and C std seems to imply it could even be something like `geany - Geany IDE` or something else representing the program name.

In any case, I still like the idea of using the _exact_ path to the desired binary where possible.

**Comment by codebrainz** [comment]
This seems to be a way to handle this for OSX bundles:
https://github.com/qtproject/qtbase/blob/1d8b6597a36f266c3384553365326cf7ec9f8641/src/corelib/kernel/qcoreapplication_mac.cpp#L43

**Comment by codebrainz** [comment]
Maybe we could do something like:

1. First try any platform-specific ways that anybody cares enough about to implement, if none or possible or fail...
2. Attempt to determine from `argv[0]`, if absolute try it directly, if not, attempt to look it up in `PATH`, if neither works, then...
3. Fall back to existing behaviour using deprecated `g_find_program_in_path()` function as it does right now (or if someone updates this code to use spawn functions and it does same), based on hard-coded name.

Sound reasonable?

**Comment by zhekov** [comment]
>> And if one execs geany, but passes an argv[0] of "foo" as exec()'s argv[0], so be it - we should try to respect that.

> Which "foo" though? The point of this PR is to try and help it use the
exact same binary code as the current instance.

> I'm a little unsure also whether argv[0] would even be reliable in all cases. According to the C standard it's up to the implementation and should "... represent the program name".

That's exactly what I meant by "foo": the incoming argv[0] can be anything that one passes to exec(), directly or with some other command. With the normal means of execution, shell or GUI, argv[0] will be one of the following:

1. Geany executable absolute name - may be used directly.
2. Geany executable relative name - basename can be combined with bindir.
3. Geany executable base name - can be combined with bindir.

What about "foo"? Personally I run Geany via "geany-bin" which sets the --socket-file based on the current desktop (there is a short example of that in the manual), so geany -i for a new window is fine *for me*. But if one uses -P or -p instead, (s)he could set argv[0] to "geany-bin" with exec -a, so that the options apply to any new window. Unfortunately, the "real executable name" solutions will disregard that.

The "session management" patch (now defunct, and the "sm" branch is even more obsolete) was much more intelligent as to which Geany CL options to keep or discard. Though even with that, a startup script may do more things than simply shuffle some options - for example, set per-instance environment variable(s) for some build tools.

My point is, if one *explicitly* sets argv[0], we should try to respect that. Otherwise, if argv[0] of the standard values, no platform code is required (except on OSX) - it's either absolute, or combinable with bindir. Thus, platform code (except OSX) is required if, and only if, one sets argv[0] to something invalid. Well, is it worth? That's not our mistake either...

**Comment by codebrainz** [comment]
I don't care enough to discuss it any more. This PR provides an improvement over the current hard-coded `"geany"` that neither respects the `argv[0]` passed to main (which is nearly undefined and could contain many things, including such stuff as the translated program name, a description of the program, a link to the documentation, some random launcher script, an empty string, etc) nor even makes an attempt to open the Geany binary currently in use. The PR also saves and then passes the configuration directory currently in use so that the new instance uses the same plugins, fonts, etc.

If anyone has more wherewithal to discuss other improvements in the comments feel free to provide other patches, for now we'll leave the worst solution in place (I can use my patch locally still :)

**Comment by techee** [comment]
@codebrainz Ah, sorry, noticed your last comment regarding the patch future just now so take the above just as a suggestion of what needs to be done in case this feature is implemented.

**Comment by zhekov** [comment]
I'm working on an argv[0] based variant, including OSX support.

**Comment by codebrainz** [comment]
I don't think we should prefer the argv[0] way over using the correct executable. I guess I can apply these changes on top of @zhekov's upcoming PR. At least both ways are better than current.

**Comment by codebrainz** [comment]
I hadn't noticed, but PR #446 covers this same topic.

**Comment by elextr** [comment]
As I suggested on IRC, maybe the simplest way would be for an entry field in preferences->tools (plenty of space there) with the default set by install to {bindir}/geany -i.  Then the user can set -c if they want to.  I for one would make it point to a *different* config to avoid the problems of overwriting if I close the instances in the wrong order.

**Comment by codebrainz** [comment]
Naw, it shouldn't be configurable, it should just "do the right thing", like:

- The exact path to the binary code of the current instance (platform-specific)
- Some guess at the path of the current based on compile-time (ex. `$(bindir)/geany`)
- Some trial and error on `argv[0]`
- Or else  just use `"geany"`

It would be a shame to admit we need a setting for this (even a 'various pref').

**Comment by elextr** [comment]
On 8 August 2015 at 13:19, Matthew Brush <notifications@github.com> wrote:

> Naw, it shouldn't be configurable, it should just "do the right thing",
> like:
>
>    - The exact path to the binary code of the current instance
>    (platform-specific)
>    - Some guess at the path of the current based on compile-time (ex.
>    $(bindir)/geany)
>
>
​well install-time, but yeah, ​curently my favourite simply because it
avoids the platform specific issues



>
>    - Some trial and error on argv[0]
>    - Or else just use "geany"
>
> It would be a shame to admit we need a setting for this (even a 'various
> pref').
>

​The above is an ok set of fallback choices for the executable to use.

But I was thinking of the options​, the -c in particular.  I can just as
easily make a use-case for no -c, -c the same as the current geany and -c
to a different config.

So I do not believe that there is an obvious "right thing" for that, and
user config is the only way.

And letting the user configure the whole command with the usual %f for the
filename is the easiest and most flexible solution.  Maybe it can also have
%x for the executable as defined by the cascade above.  Also %c for the
config used by current Geany.

Then the default just needs to be `%x -i -c %c %f` to get the current
behaviour, but with the right executable.

**Comment by codebrainz** [comment]
On 15-08-07 08:39 PM, elextr wrote:
> On 8 August 2015 at 13:19, Matthew Brush <notifications@github.com> wrote:
>
>> Naw, it shouldn't be configurable, it should just "do the right thing",
>> like:
>>
>>     - The exact path to the binary code of the current instance
>>     (platform-specific)
>>     - Some guess at the path of the current based on compile-time (ex.
>>     $(bindir)/geany)
>>
>>
> ​well install-time, but yeah, ​curently my favourite simply because it
> avoids the platform specific issues
>
>
>
>>
>>     - Some trial and error on argv[0]
>>     - Or else just use "geany"
>>
>> It would be a shame to admit we need a setting for this (even a 'various
>> pref').
>>
>
> ​The above is an ok set of fallback choices for the executable to use.
>
> But I was thinking of the options​, the -c in particular.  I can just as
> easily make a use-case for no -c, -c the same as the current geany and -c
> to a different config.
>

I'd be curious to hear that for the "Open in New Window" feature which 
in most applications would just open another top-level window.

> So I do not believe that there is an obvious "right thing" for that, and
> user config is the only way.
>
> And letting the user configure the whole command with the usual %f for the
> filename is the easiest and most flexible solution.  Maybe it can also have
> %x for the executable as defined by the cascade above.  Also %c for the
> config used by current Geany.
>
> Then the default just needs to be `%x -i -c %c %f` to get the current
> behaviour, but with the right executable.
>

I think this would be fine as a "various preference", to fallback or 
override. If you really wanted to get complicated, you could use `%s` 
for "self exe", `%a` for `argv[0]`, etc.

For me, I don't care as long as the new instance is the same exe and 
config dir as the current instance.

**Comment by zhekov** [comment]
I wrote my "Open in New Window" variant, and would like to discuss it here before publishing, if @codebrainz does not object.

\1. On startup, it initializes app->executable with:

- osx_bundle: gtkosx_application_get_executable_path()

Reason: It's unclear to me whether restarting a bundled executable via script is permissible, there seem to be some limitations. I don't know enough about OSX.

- win32: win32_get_executable_path()

Reason: Can't restart via script, argv[0] is always the real Geany excutable.

- unix (not osx bundle), or failed win32: tm_get_real_path(argv[0]) + g_file_test(G_FILE_TEST_IS_EXECUTABLE), on failure -> g_find_program_in_path(argv[0]), on failure -> "geany".

Reason: avoid testing for X unix paths and Y syscalls, support startup script. Though I reserved a place to put system-dependent unix code if needed.

\2. On new instance, it checks if app->executable is NULL, which may only mean failed gtkosx_application_get_executable_path(). If so, Geany is started as a bundle, with added "--args", "-i", "-c", app->configdir if OSX >= 10.6; otherwise, it's business as usual.

Now, given the relationship between a bundle and it's executable, I can't imagine a situation when OSX get_resource_path() will work, but get_executable_path() will not. And I'm not aware of any reason GetModuleFileName(NULL, buffer, size >= MAX_PATH) will fail, but the process as such will be fine.

**Comment by elextr** [comment]
On 8 August 2015 at 21:15, zhekov <notifications@github.com> wrote:

> I wrote my "Open in New Window" variant, and would like to discuss it here
> before publishing, if @codebrainz <https://github.com/codebrainz> does
> not object.
>
> \1. On startup, it initializes app->executable with:
>
>    - osx_bundle: gtkosx_application_get_executable_path()
>
> Reason: It's unclear to me whether restarting a bundled executable via
> script is permissible, there seem to be some limitations. I don't know
> enough about OSX.
>
>    - win32: win32_get_executable_path()
>
> Reason: Can't restart via script, argv[0] is always the real Geany
> excutable.
>
>    - unix (not osx bundle), or failed win32: tm_get_real_path(argv[0]) +
>
>
​If argv[0] is not an absolute path the executable will have been found by
searching PATH, it will not necessarily be the absolute version of argv[0]
which depends on the CWD, but the CWD may contain a geany executable and so
this test will give the wrong answer.​

**Comment by zhekov** [comment]
> If argv[0] is not an absolute path the executable will have been found by searching PATH

Actually, only base names are searched in the path...

> but the CWD may contain a geany executable and so this test will give the wrong answer

...but you are right that a base name in CWD will be wrong. And since g_find_program_in_path is deprecated for some reason, I'd better rewrite this thing.

**Comment by elextr** [comment]
On 10 August 2015 at 02:05, zhekov <notifications@github.com> wrote:

> If argv[0] is not an absolute path the executable will have been found by
> searching PATH
>
> Actually, only base names are searched in the path...
>

​Sorry, thats what I meant to say.
​

**Comment by b4n** [comment]
> The point of this PR is to try and help it use the exact same binary code as the current instance.

FWIW, this is just not possible (AFAIK), as the executable might have been modified since current instance was launched.  To really use the same binary, one would probably have to `fork()` without `exec()`ing afterward -- but this has other implications.

As of my opinion on this whole issue… I don't know.  No, current situation is not good, as it'll fail in relatively common situations (if Geany wasn't installed in `$PATH`).  But I also believe the complexity should be kept at a reasonable minimum, it doesn't seem to be a critical enough problem to warrant "lots" of platform-specific code we'll have to maintain.
Before the discussion here, I'd have went with a variant of #446 (with fixed OSX bundle support, I think I even have the commit somewhere), e.g. just spawn `$bindir/geany`.  However I now realize that even this might be incorrect, as the executable name can be transformed (`--program-prefix`/`--program-suffix`/`--program-transform-name`) -- though, how important is it?

I also understand @zhekov's use case of a startup script, which the user would probably want to see used when spawning the additional instance.  And to support that, I only see two solutions: @zhekov's of trying `argv[0]`, and @elextr's of using a configurable command.

Anyhow, I don't mind much whatever solution is proposed in the end, but I'd really like it to be as platform-independent and small as possible, as again, I don't think this feature is critical enough to warrant lot of code for slight a gain (e.g. I'd rather see it easily maintainable than totally accurate).

**Comment by techee** [comment]
By the way when using the bundle under OS X there's always a launcher script used to run the binary - it has to set up the correct paths to GTK inside the bundle and set some additional environment variables. So using argv[0] wouldn't work here.

I agree with @b4n that some minimalistic approach should be taken to avoid adding too much code and UI preferences which are basically never used. #446 looks good enough to me.

**Comment by codebrainz** [comment]
@techee if there's a launcher script, then will any way other than `open Geany` ever actually work with the bundle?

**Comment by codebrainz** [comment]
As nobody had commented on the commits yet and I had a couple commits fixing issues Travis found, I rebased and forced pushed this branch to make it the single commit it should've been.

**Comment by techee** [comment]
> if there's a launcher script, then will any way other than open Geany ever actually work with the bundle?

@codebrainz The gtkosx_application_get_executable_path() returns the path of the "primary" executable, which is the launcher script. The bundle-creation script puts the launcher script under

```
/Applications/Geany.app/Contents/MacOS/geany
```

while the binary itself is

```
/Applications/Geany.app/Contents/MacOS/geany-bin
```

which is then executed by the script.

**Comment by kugel-** [comment]
I was going to suggest that trying too hard is hopeless effort. Thus I agree with the spirit of fd65e82. I can see a problem for portable/relocatable installations too, though

File: src/utils.c [hunk 1]
```diff
@@ -2152,39 +2153,56 @@ const gchar *utils_resource_dir(GeanyResourceDirType type)
...
+	if (is_osx_bundle())
+	{
+		gchar *exec_file = gtkosx_application_get_executable_path();
+		if (exec_file != NULL)
+			return exe_file;
+	}
+#endif
+
+	// Try using the installation directory
+	gchar *exe_file = g_build_filename(GEANY_PREFIX, "bin", "geany", NULL);
```
Comment (techee) [inline src/utils.c]: You could declare the exe_file variable at the beginning of the function and use it both for the OS X and non-OS X case.

**Comment by codebrainz** [comment]
Closing since I'm not working on this anymore, and various other PRs such as #446 and #653 provide overlapping improvements. I'll re-open if I ever want to pickup working on this topic.

**Pull request closed without merging**